elements/bulbs-pinned-element/bulbs-pinned-element.js

```diff
@@ -1,128 +1,150 @@
-import { registerElement, BulbsHTMLElement } from 'bulbs-elements/register';
 import invariant from 'invariant';
-import { getScrollOffset, InViewMonitor } from 'bulbs-elements/util';
+
+import {
+  registerElement,
+  BulbsHTMLElement
+} from 'bulbs-elements/register';
+import {
+  InViewMonitor,
+  getRoundedBoundingClientRect,
+  getScrollOffset,
+  moveChildren
+} from 'bulbs-elements/util';
 import './bulbs-pinned-element.scss';
 
 export default class BulbsPinnedElement extends BulbsHTMLElement {
+
   attachedCallback () {
-    invariant(this.hasAttribute('pinned-to'), '<bulbs-pinned-element pinned-to=".selector">: a pinned-to selector is required');
 
-    const selector = this.getAttribute('pinned-to');
-    const element = document.querySelector(selector);
+    this.topOffsetAdjustment = parseInt(this.getAttribute('offset-top-px') || 0, 10);
     this.lastPosition = 0;
+    this.lastRailHeight = 0;
     this.animationRequest = null;
 
-    invariant(element, `<bulbs-pinned-element pinned-to=".selector">: no element with the selector "${selector}" is in the DOM`);
+    let car = this.querySelector('bulbs-pinned-element-car');
+    if (car) {
+      this.car = car;
+    } else {
+      this.car = document.createElement('bulbs-pinned-element-car');
+      moveChildren(this, this.car);
+      this.appendChild(this.car);
+    }
+
+    this.boundPositionCar = this.positionCar.bind(this);
+
+    window.addEventListener('scroll', this.boundPositionCar);
+
+    this.boundPositionCar();
+  }
+
+  detachedCallback () {
+    window.removeEventListener('scroll', this.boundPositionCar);
+  }
+
+  getBoundingRects () {
+
+    return {
+      car: getRoundedBoundingClientRect(this.car),
+      parent: getRoundedBoundingClientRect(this.parentElement),
+      rail: getRoundedBoundingClientRect(this),
+    };
+  }
 
-    window.addEventListener('scroll', this.handleScrollEvent.bind(this));
+  isInView (boundingRects) {
+    return InViewMonitor.isElementInViewport(this, boundingRects.rail);
   }
 
-  handleScrollEvent () {
+  positionCar () {
     if(!this.animationRequest) {
       this.animationRequest = requestAnimationFrame(() => {
         this.animationRequest = null;
 
-        if(!InViewMonitor.isElementInViewport(this, this.getBoundingClientRect())) { return; }
+        const boundingRects = this.getBoundingRects();
 
-        let elementPinnedTo = this.getElementPinnedTo();
-        const boundingRects = this.getBoundingRects(elementPinnedTo);
+        this.style.height = `${boundingRects.parent.height - Math.abs(boundingRects.parent.top - boundingRects.rail.top)}px`;
+        this.style.width = `${boundingRects.parent.width}px`;
 
-        if(this.isScrollingDown()) {
-          this.handleScrollDown(elementPinnedTo, boundingRects);
+        if (this.hasNewRailHeight(boundingRects)) {
```

Comment:
Allows adjustments when the rail changes heights, i.e. new content has loaded into a reading list and now the sidebar is taller.